src/sdformat_urdf.cpp

```diff
@@ -0,0 +1,466 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "sdformat_urdf/sdformat_urdf.hpp"
+
+#include <algorithm>
+#include <string>
+#include <vector>
+
+#include <resource_retriever/retriever.h>
+#include <sdf/sdf.hh>
+#include <urdf_world/types.h>
+#include <urdf_model/model.h>
+
+namespace sdformat_urdf {
+/// \brief Convert SDFormat Link to URDF Link
+urdf::LinkSharedPtr
+convert_link(const sdf::Link & sdf_link, sdf::Errors & errors);
+
+/// \brief Convert SDFormat Joint to URDF Joint
+urdf::JointSharedPtr
+convert_joint(const sdf::Joint & sdf_joint, sdf::Errors & errors);
+
+urdf::Pose
+convert_pose(const ignition::math::Pose3d & sdf_pose);
+
+urdf::GeometrySharedPtr
+convert_geometry(const sdf::Geometry & sdf_geometry, sdf::Errors & errors);
+}  // namespace sdformat_urdf
+
+urdf::ModelInterfaceSharedPtr
+sdformat_urdf::parse(const std::string & data, sdf::Errors & errors)
+{
+  auto sdf_dom = std::make_shared<sdf::Root>();
+  errors = sdf_dom->LoadSdfString(data);
+  if (sdf_dom) {
+    return sdformat_urdf::sdf_to_urdf(*sdf_dom, errors);
+  }
+  return nullptr;
+}
+
+
+urdf::ModelInterfaceSharedPtr
+sdformat_urdf::sdf_to_urdf(const sdf::Root & sdf_dom, sdf::Errors & errors)
+{
+  if (sdf_dom.WorldCount()) {
+    errors.emplace_back(
+      sdf::ErrorCode::STRING_READ,
+      "SDFormat xml has a world; but only a single model is supported");
+    return nullptr;
+  }
+  if (1u != sdf_dom.ModelCount()) {
+    errors.emplace_back(
+      sdf::ErrorCode::STRING_READ,
+      "SDFormat xml has multiple models; but only a single model is supported");
+    return nullptr;
+  }
+
+  return convert_model(*sdf_dom.ModelByIndex(0), errors);
+}
+
+urdf::ModelInterfaceSharedPtr
+sdformat_urdf::convert_model(const sdf::Model & sdf_model, sdf::Errors & errors)
+{
+  urdf::ModelInterfaceSharedPtr urdf_model = std::make_shared<urdf::ModelInterface>();
+
+  // copy name
+  urdf_model->name_ = sdf_model.Name();
+
+  // create matching links
+  for (uint64_t l = 0; l < sdf_model.LinkCount(); ++l) {
+    const sdf::Link * sdf_link = sdf_model.LinkByIndex(l);
+
+    if (!sdf_link) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to get sdf link");
+      return nullptr;
+    }
+
+    auto pair = urdf_model->links_.emplace(sdf_link->Name(), convert_link(*sdf_link, errors));
+
+    if (!pair.second) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to create element in links map");
+      return nullptr;
+    }
+
+    const std::shared_ptr<urdf::Link> & urdf_link = pair.first->second;
+
+    if (!urdf_link) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to convert sdf::link [" + sdf_link->Name() + "] to urdf::Link");
+      return nullptr;
+    }
+  }
+
+  // sdf canonical link -> urdf root link
```

Comment:
this may be true for specific models but is not true in general. I think we could add some helper APIs to libsdformat to aid in determining:

* if the directed graph in an SDFormat file is also a directed tree
* and if so, which is the root link of that directed tree

we are brainstorming what those APIs could look like in https://github.com/osrf/sdformat/issues/273#issuecomment-663292059

for now, this assumption is fine

Reply:
Perhaps this can be worded as a `TODO` pointing to that issue? e.g.
```
TODO(name) Possibly relax this assumption given the outcome of osrf/sdformat#273.
```

Reply:
Reworded assumption + added TODO c84a46db49923115a56319d0ef0f07a6136b97d1